pkg/util/containers/v2/metrics/provider/cache.go

```diff
@@ -0,0 +1,71 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+package provider
+
+import (
+	"sync"
+	"time"
+)
+
+type cacheEntry struct {
+	value     interface{}
+	err       error
+	timestamp time.Time
+}
+
+// Cache provides a caching mechanism based on staleness toleration provided by requestor
+type Cache struct {
+	cache       map[string]cacheEntry
+	cacheLock   sync.RWMutex
+	gcInterval  time.Duration
+	gcTimestamp time.Time
+}
+
+// NewCache returns a new cache dedicated to a collector
+func NewCache(gcInterval time.Duration) *Cache {
+	return &Cache{
+		cache:      make(map[string]cacheEntry),
+		gcInterval: gcInterval,
+	}
+}
+
+// Get retrieves data from cache, returns not found if cacheValidity == 0
+func (c *Cache) Get(currentTime time.Time, key string, cacheValidity time.Duration) (interface{}, bool, error) {
+	if cacheValidity <= 0 {
+		return nil, false, nil
+	}
+
+	c.cacheLock.RLock()
+	entry, found := c.cache[key]
+	c.cacheLock.RUnlock()
+
+	if !found {
+		return nil, false, nil
+	}
+
+	if currentTime.Sub(entry.timestamp) > cacheValidity {
+		return nil, false, nil
+	}
+
+	if entry.err != nil {
+		return nil, true, entry.err
+	}
+
+	return entry.value, true, nil
+}
+
+// Store retrieves data from cache
+func (c *Cache) Store(currentTime time.Time, key string, value interface{}, err error) {
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+
+	if currentTime.Sub(c.gcTimestamp) > c.gcInterval {
```

Comment:
... and if the other cache can't be used: this looks like it may expire fresh keys if they were just created, but `gcInterval` was almost elapsed. If the intention is to avoid a separate goroutine to handle the expiration, I'd suggest `cache.Cache` a bit harder ;)